pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go

```diff
@@ -0,0 +1,343 @@
+			}
+		}
+
+		volumeID, err := util.GetVolumeID(ctx, r.client, instance.Spec.PvcName, instance.Namespace)
+		if err != nil {
+			msg := fmt.Sprintf("Failed to get volumeID from pvcName: %q. Error: %+v", instance.Spec.PvcName, err)
+			log.Error(msg)
+			setInstanceError(ctx, r, instance, msg)
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+
+		// Query volume
+		log.Infof("Querying volume: %s for CnsFileAccessConfig request with name: %q on namespace: %q",
+			volumeID, instance.Name, instance.Namespace)
+		volume, err := common.QueryVolumeByID(ctx, r.volumeManager, volumeID)
+		if err != nil {
+			if err.Error() == common.ErrNotFound.Error() {
+				msg := fmt.Sprintf("CNS Volume: %s not found", volumeID)
+				log.Error(msg)
+				setInstanceError(ctx, r, instance, msg)
+				return reconcile.Result{RequeueAfter: timeout}, nil
+			}
+			log.Errorf("Failed to query CNS volume: %s with error: %+v", volumeID, err)
+			setInstanceError(ctx, r, instance, "Unable to find the volume in CNS")
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+
+		vSANFileBackingDetails := volume.BackingObjectDetails.(*cnstypes.CnsVsanFileShareBackingDetails)
+		accessPoints := make(map[string]string)
+		for _, kv := range vSANFileBackingDetails.AccessPoints {
+			accessPoints[kv.Key] = kv.Value
+		}
+		if len(accessPoints) == 0 {
+			log.Errorf("No access points found for volume: %q", volumeID)
+			setInstanceError(ctx, r, instance, "No access points found.")
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+
+		// Read the CnsFileAccessConfig instance again because the instance is already modified
+		err = r.client.Get(ctx, request.NamespacedName, instance)
+		if err != nil {
+			msg := fmt.Sprintf("Error reading the CnsFileAccessConfig with name: %q on namespace: %q. Err: %+v",
+				request.Name, request.Namespace, err)
+			// Error reading the object - requeue the request.
+			recordEvent(ctx, r, instance, v1.EventTypeWarning, msg)
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+
+		// Update the instance to indicate the volume registration is successful
+		msg := fmt.Sprintf("Successfully configured access points of VM: %q on the volume: %q", instance.Spec.VMName, instance.Spec.PvcName)
+		instance.Status.AccessPoints = accessPoints
+		err = setInstanceSuccess(ctx, r, instance, msg)
+		if err != nil {
+			msg := fmt.Sprintf("Failed to update CnsFileAccessConfig instance with error: %+v", err)
+			log.Error(msg)
+			setInstanceError(ctx, r, instance, msg)
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+		log.Info(msg)
+	}
+
+	if instance.DeletionTimestamp != nil {
+		removeFinalizerFromCRDInstance(ctx, instance)
+		return reconcile.Result{}, nil
+	}
+
+	backOffDurationMapMutex.Lock()
+	delete(backOffDuration, instance.Name)
+	backOffDurationMapMutex.Unlock()
+	return reconcile.Result{}, nil
+}
+
+// setInstanceSuccess sets instance to success and records an event on the CnsRegisterVolume instance
```

Comment:
Cant all the functions from here on be moved to `cnsoperator/util/util.go`? I see that other controllers use the same logic, the only difference is the `instance` type. Have you tried using `runtime.Object` as a parameter instead of the exact type, and then cast it accordingly? All controllers can then leverage the same functions.

Reply:
Created this issue: https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/587 for tracking the suggestion.
I am not addressing the review comment in this PR to avoid more complexity/testing cycles. It would need some additional effort to refactor 4 different Custom resource controller.go files and also run some tests.
Thanks

Reply:
CnsRegisterVolume -> CnsFileAccessConfig in method doc.